Fetch service alerts, filter them and include in route-pt responses

graue opened the pull request:
When the GraphHopper `route-pt` endpoint is hit, this fetches service alerts from 511.org (or in theory whatever provider is configured), filters them to only those alerts that apply to the returned routes, and adds them to the response returned by GraphHopper in the `service_alerts` key.

Mostly done. Still to do as of 3/11, before merging:

- [ ] ~needs to handle alerts with no agency_id specified (I just realized that is possible)~
- [ ] ~needs to handle alerts with a direction_id or trip~
- [x] needs caching fixed to not make multiple simultaneous requests

3/14: I've changed my mind about handling the direction-ID and no-agency-ID cases. No Bay Area transit provider is currently using them, and I think if not supporting those edge cases turns out to be a problem later, it makes more sense to fix it once UI is implemented so we can see what alerts we're looking at.

Andykmcc commented on src/config.js:
```diff
@@ -6,3 +10,40 @@ export const HOSTNAME = process.env.HOSTNAME;
+  );
+}
+
+const dotenvPath = '.env';
```
this all seems needless complex. i've never done any of this when using dotenv. why not just do the standard thing:
```javascript
import * as dotenv from 'dotenv';
dotenv.config();
```

Andykmcc commented on src/config.js:
```diff
@@ -6,3 +10,40 @@ export const HOSTNAME = process.env.HOSTNAME;
+  NODE_ENV !== 'test' && `${dotenvPath}.local`,
+  `${dotenvPath}.${NODE_ENV}`,
+  dotenvPath,
+].filter(Boolean);
```
can you change this `.filter(Boolean)` to something more idiomatic? It looks like it filters all Boolean values but it actually filters all false-y values.

Andykmcc commented on src/config.js:
```diff
@@ -1,3 +1,7 @@
+import fs from 'fs';
+import * as dotenv from 'dotenv';
+import * as dotenvExpand from 'dotenv-expand';
```
we you really need this? I dont see dotenv-expand featured used in any of the committed files. it add complexity to this project and enabled more complex patterns that I don't think we should have in env vars. KISS.

Andykmcc commented on src/config.js:
```diff
@@ -6,3 +10,40 @@ export const HOSTNAME = process.env.HOSTNAME;
+  // Don't include `.env.local` for `test` environment
+  // since normally you expect tests to produce the same
+  // results for everyone
+  NODE_ENV !== 'test' && `${dotenvPath}.local`,
```
I don think we should do this. right now we dont need the test file but in other projects i've needed a .env.test file to set env vars that are useful in unit testing. I think if such a file exists we should commit it and use it.

Andykmcc commented on src/graphhopper/router.js:
```diff
@@ -1,9 +1,50 @@
-const routes = [
+router.use((req, res, next) => {
+  // you only want to cache for GET requests
+  if (req.method == 'GET') {
```
`===`

Andykmcc commented on src/graphhopper/router.js:
```diff
@@ -1,9 +1,50 @@
+    if (graphHopperResult.status === 'rejected') throw graphHopperResult.reason;
+    if (alertResult.status === 'rejected') logger.error(alertResult.reason);
+
+    res.send(mergeAlertsIntoRoutes(alertResult.value, graphHopperResult.value.data));
```
```suggestion
    res.json(mergeAlertsIntoRoutes(alertResult.value, graphHopperResult.value.data));
```

graue replied:
Can you explain why you're asking for this change? What is wrong with the current way of doing it?

The use of `res.send()` is unchanged from your code on main. I am not familiar enough with this API to know what the difference is between calling `.send()` or `.json()`, and have chosen to make the minimum possible change to the code to accomplish my goal. If I change it, then I have to go research the API to make sure the way that you're preferring now will work and not break anything. That's significant added work and I'm not clear on what the benefit is.

Andykmcc commented on src/graphhopper/router.js:
```diff
@@ -1,9 +1,50 @@
+    else {
+      res.status(500);
+    }
+    res.end();
```
move this to a `finally` block. as in:
```javascript
try {...}
catch {...}
finally { res.end() }
```

graue replied:
Same question as above -- why is that better? What's the difference? As above, this usage is unchanged from the code on main that you wrote.

Andykmcc commented on src/index.js:
```diff
@@ -11,10 +11,13 @@ import { router as photonRouter } from './photon/index.js';
 import { router as nominatimRouter } from './nominatim/index.js';
 import { router as fileRouter } from './file/index.js';
 
+if (!process.version.startsWith('v18.')) {
```
I dont really like this but I think I get why you want it. id rather this be gone or at the very least try to match `>=18` since node versions change so rapidly.

Andykmcc commented on src/gtfs-rt/alerts.js:
```diff
@@ -0,0 +1,144 @@
+    }
+  }
+
+  const relevantAlerts = alerts.filter(alert => {
```
thanks for figuring out all these rules. 🙌 . personally I prefer a bunch of small functions then a chain of `.filter()`s that applies each of them separately. This mostly come from wanting to make unit testing cleaner/simpler. at some point I'll add tests to this repo at which point I will likely refactor this a bit. but we can discuss it then if I ever get around to it.

Andykmcc commented on src/gtfs-rt/client.js:
```diff
@@ -0,0 +1,72 @@
+import GtfsRealtimeBindings from 'gtfs-realtime-bindings';
+import logger from '../lib/logger.js';
+
+const API_KEY = process.env.GTFS_REALTIME_TOKEN;
```
can you move these `process.env` refs into the src/config.js file, then import that here? My goal with that file was to keep all env var refs confined to there for simply reference and to make the eventual shift to dependency injection easier.

graue replied:
done

Andykmcc commented on src/gtfs-rt/client.js:
```diff
@@ -0,0 +1,72 @@
+let _alertsPromise;
+// We're allowed to make 60 requests per hour, let's be conservative and
+// cache for 3min to make it a maximum of 20
+const CACHE_LENGTH = 3 * 60 * 1000;
```
can you move this to env env var so we can tweak it as needed without needing to do a new build.

graue replied:
done

Andykmcc commented on src/gtfs-rt/client.js:
```diff
@@ -0,0 +1,72 @@
+
+  // If there's already a fetch in progress, we'll wait for that one.
+  // But if not, let's initiate one.
+  if (!_alertsPromise)
```
this will not have the effect you are describing because this file isn't a singleton. each time it is imported it will create a new `_alertsPromise` for that context. assuming this file is only ever used  in one import statement then it will work as described but this comment/guard suggests it provides a level of protection that it really doesn't. I think the idea though, sadly without a DI system producing singletons is kind of a pain in the ass in JS. maybe we could just be more explicit in the comment about the depth of this gaurd?

graue replied:
Hmm I don't think this is a problem in practice or else using `instanceof` within Node.js wouldn't work.

[This Medium post from 2016](https://medium.com/@lazlojuly/are-node-js-modules-singletons-764ae97519af) suggests that while there are some cases in which (old) Node.js would not cache (CommonJS) modules, we would be unlikely to hit those cases in practice.

[Current Node's documentation on ES modules](https://nodejs.org/api/esm.html) is unclear on the point.

[Here's a free ebook chapter](https://exploringjs.com/es6/ch_modules.html#sec_modules-in-javascript) that says of ES modules, "Modules are singletons. Even if a module is imported multiple times, only a single “instance” of it exists." (Found via [this StackOverflow answer](https://stackoverflow.com/questions/64477263/use-module-as-singleton), which makes the point about `instanceof`.)

Andykmcc reviewed (changes requested):
❤️

graue commented:
All the dotenv stuff is copied over from bikehopper-ui (config/env.js). We may not need all of it, but I wanted to make it work the same between both projects to prevent surprises. Ideally both bikehopper-ui and bikehopper-web-app will be combined into a single project at some point.

graue commented:
@Andykmcc, I've addressed your comments as follows:
- the references to env vars are all in src/config.js
- cache time is configurable as an env var
- Node version check uses >=18
- use === instead of == (this is actually unchanged from the existing code on main, but I'm happy to correct since the change is trivial and obviously correct)

Other than that, I prefer not to tweak existing code unrelated to my change, and I prefer to keep the handling of `.env.*` files identical to what it is in bikehopper-ui to prevent unpleasant surprises where the two repos work subtly differently. Let me know if that's ok with you.
